File: Algorithms/src/2948. Make Lexicographically Smallest Array by Swapping Elements/Code/solution.py
from typing import List

class Solution:
    def lexicographicallySmallestArray(self, nums: List[int], limit: int) -> List[int]:
        n = len(nums)  # Store the size of the array.
        
        # Store each value together with its original index.
        elements = [(nums[i], i) for i in range(n)]
        
        # Sort by value so connected values become consecutive.
        elements.sort()
        
        # Store the lexicographically smallest result.
        answer = [0] * n
        
        start = 0  # First element of the current connected group.
        
        while start < n:
            end = start  # Expand the current group.
            
            # Consecutive values belong to the same group when their
            # difference is at most limit.
            while (
                end + 1 < n
                and elements[end + 1][0] - elements[end][0] <= limit
            ):
                end += 1
            
            # Collect all original indices belonging to this group.
            indices = []
            
            for i in range(start, end + 1):
                indices.append(elements[i][1])
            
            # Sort positions so smaller values go to earlier positions.
            indices.sort()
            
            # The values are already sorted because elements was sorted.
            for i, index in enumerate(indices):
                answer[index] = elements[start + i][0]
            
            # Move to the next connected group.
            start = end + 1
        
        return answer  # Return the final lexicographically smallest array.

File: Algorithms/src/2948. Make Lexicographically Smallest Array by Swapping Elements/Code/solution.java
import java.util.*;

class Solution {
    public int[] lexicographicallySmallestArray(int[] nums, int limit) {
        int n = nums.length; // Store the array size for easier use.
        
        // Each pair stores {value, originalIndex}.
        int[][] elements = new int[n][2];
        
        for (int i = 0; i < n; i++) {
            elements[i][0] = nums[i]; // Store the value.
            elements[i][1] = i;       // Store its original position.
        }
        
        // Sort all pairs by their values.
        Arrays.sort(elements, (a, b) -> Integer.compare(a[0], b[0]));
        
        // Store the final lexicographically smallest array.
        int[] answer = new int[n];
        
        int start = 0; // First position of the current connected group.
        
        while (start < n) {
            int end = start; // Expand the current group.
            
            // Keep consecutive values in the same group while they can
            // be connected through valid swaps.
            while (end + 1 < n &&
                   (long) elements[end + 1][0] - elements[end][0] <= limit) {
                end++;
            }
            
            // Store all original indices belonging to this group.
            int[] indices = new int[end - start + 1];
            
            for (int i = start; i <= end; i++) {
                indices[i - start] = elements[i][1];
            }
            
            // Sort positions so smaller values go to earlier indices.
            Arrays.sort(indices);
            
            // The values in the current group are already sorted because
            // the complete elements array was sorted by value.
            for (int i = 0; i < indices.length; i++) {
                answer[indices[i]] = elements[start + i][0];
            }
            
            // Move to the next connected group.
            start = end + 1;
        }
        
        return answer; // Return the final array.
    }
}